### PR TITLE
Feature swxsupport

### PR DESCRIPTION
**Pull request by ilkkarinne**
TAC-projektin SWX-tuki

### CODE REVIEW

**Review by petringo — APPROVED** [review]
I think this is now mature enough for merging. The commented parts should be revised, but won't prevent proceeding now with the merge. In addition, it might be a good idea to check IDE inspections for ways to do some minor code cleaning.

File: src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/SWXEffect.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+public class SWXEffect extends RegexMatchingLexemeVisitor {
+    public SWXEffect(final OccurrenceFrequency prio) {
+        super("^(?<phenomenon>(SATCOM|HF\\sCOM|GNSS|RADIATION){1}\\s(MOD|SEV){1})$", prio);
+    }
+
+    @Override
+    public void visitIfMatched(final Lexeme token, final Matcher match, final ConversionHints hints) {
+        token.identify(LexemeIdentity.SWX_EFFECT);
+        SpaceWeatherPhenomenon phenomenon = SpaceWeatherPhenomenonImpl.builder().fromCombinedCode(match.group("phenomenon")).build();
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/SWXEffect.java:27]: This comment has been adressed

[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/SWXEffect.java:27]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+            final List<String> remarks = getRemarks(match, hints);
+            if (!remarks.isEmpty()) {
+                builder.setRemarks(remarks);
+            }
+        });
+
+
+        retval.addIssue(conversionIssues);
+        if (conversionIssues.size() == 0) {
+            retval.setConvertedMessage(builder.build());
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+                        + "unit for airspace volume"));
+            }
+            verticalLimitOperator = Optional.ofNullable(
+                    l.getParsedValue(Lexeme.ParsedValueName.RELATIONAL_OPERATOR, AviationCodeListUser.RelationalOperator.class));
+        }
+
+        l = lexeme.findNext(LexemeIdentity.POLYGON_COORDINATE_PAIR);
+        if (l != null) {
+            final PolygonGeometryImpl.Builder polyBuilder = PolygonGeometryImpl.builder()//
+                    .setSrsName("http://www.opengis.net/def/crs/EPSG/0/4326")//
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java:242]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                        //Upper right corner:
+                        location.get().getLatitudeBandMinCoordinate().ifPresent(coordinates::add);
+                        coordinates.add(maxLongitude.orElse(180d));
+
+                        //Upper left corner (again, to close the ring):
+                        location.get().getLatitudeBandMinCoordinate().ifPresent(coordinates::add);
+                        coordinates.add(minLongitude.orElse(-180d));
+                        final PolygonGeometry polygon = PolygonGeometryImpl.builder()
+                                .addAllExteriorRingPositions(coordinates)
+                                .setSrsName("http://www.opengis.net/def/crs/EPSG/0/4326")
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java:309]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/AdvisoryNumber.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import fi.fmi.avi.converter.tac.lexer.impl.FactoryBasedReconstructor;
+import fi.fmi.avi.converter.tac.lexer.impl.ReconstructorContext;
+import fi.fmi.avi.converter.tac.lexer.impl.RegexMatchingLexemeVisitor;
+import fi.fmi.avi.model.AviationWeatherMessageOrCollection;
+import fi.fmi.avi.model.swx.SpaceWeatherAdvisory;
+import fi.fmi.avi.model.swx.immutable.AdvisoryNumberImpl;
+
+public class AdvisoryNumber extends RegexMatchingLexemeVisitor {
+    public AdvisoryNumber(final OccurrenceFrequency prio) {
+        super("^(?<advisoryNumber>[\\d]{4}/[0-9][0-9]?[0-9]?[0-9]?)$", prio);
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/AdvisoryNumber.java:19]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/ReplaceAdvisoryNumber.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import fi.fmi.avi.converter.tac.lexer.impl.FactoryBasedReconstructor;
+import fi.fmi.avi.converter.tac.lexer.impl.ReconstructorContext;
+import fi.fmi.avi.converter.tac.lexer.impl.RegexMatchingLexemeVisitor;
+import fi.fmi.avi.model.AviationWeatherMessageOrCollection;
+import fi.fmi.avi.model.swx.SpaceWeatherAdvisory;
+import fi.fmi.avi.model.swx.immutable.AdvisoryNumberImpl;
+
+public class ReplaceAdvisoryNumber extends RegexMatchingLexemeVisitor {
+    public ReplaceAdvisoryNumber(final OccurrenceFrequency prio) {
+        super("^(?<advisoryNumber>[\\d]{4}/[0-9][0-9]?[0-9]?[0-9]?)$", prio);
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/lexer/impl/token/ReplaceAdvisoryNumber.java:19]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+            if (!remarks.isEmpty()) {
+                builder.setRemarks(remarks);
+            }
+        });
+
+
+        retval.addIssue(conversionIssues);
+        if (conversionIssues.size() == 0) {
+            builder.addAllAnalyses(analyses);
+            retval.setConvertedMessage(builder.build());
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java:153]: This comment has been addressed

File: src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+                    "Message does not start with a start token: " + firstLexeme.getTACToken()));
+        }
+
+        SpaceWeatherAdvisoryImpl.Builder builder = SpaceWeatherAdvisoryImpl.builder();
+
+        if (lexed.getTAC() != null) {
+            builder.setTranslatedTAC(lexed.getTAC());
+        }
+
+        firstLexeme.findNext(LexemeIdentity.TEST_OR_EXCERCISE_LABEL, (match) -> {
```
[the remark this answers is not in the document]
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/swx/SWXTACParser.java:76]: This comment is adressed in pr (https://github.com/fmidev/fmi-avi-messageconverter-tac/pull/56) for issue #46

**Pull request merged**